proposals/generic-attributes.md

```diff
@@ -0,0 +1,66 @@
+# Generic Attributes
+
+* [x] Proposed
+* [x] Prototype: [N/A](#)
+* [x] Implementation: [In Progress](https://github.com/dotnet/roslyn/tree/features/generic-attributes)
+* [ ] Specification: [Not Started](#)
+
+## Summary
+[summary]: #summary
+
+When generics were introduced in C# 2.0, attribute classes were not allowed to participate. We can make the language more composable by removing (rather, loosening) this restriction. The .NET Core runtime has added support for generic attributes. Now, all that's missing is support for generic attributes in the compiler.
+
+## Motivation
+[motivation]: #motivation
+
+Currently attribute authors can take a `System.Type` as a parameter and have users pass a `typeof` expression to provide the attribute with types that it needs. However, outside of analyzers, there's no way for an attribute author to constrain what types are allowed to be passed to an attribute via `typeof`. If attributes could be generic, then attribute authors could use the existing system of type parameter constraints to express the requirements for the types they take as input.
+
+## Detailed design
+[design]: #detailed-design
+
+The following section is amended: https://github.com/dotnet/csharplang/blob/main/spec/classes.md#base-classes
+
+> The direct base class of a class type must not be any of the following types: System.Array, System.Delegate, System.MulticastDelegate, System.Enum, or System.ValueType. ~~Furthermore, a generic class declaration cannot use System.Attribute as a direct or indirect base class.~~
+
+One important note is that the following section of the spec is *unaffected* when referencing the point of usage of an attribute, i.e. within an attribute list: https://github.com/dotnet/csharplang/blob/main/spec/types.md#type-parameters
+
+> A type parameter cannot be used anywhere within an attribute.
+
+This means that when a generic attribute is used, its construction needs to be fully "closed", i.e. not containing any type parameters, which means the following is still disallowed:
```

Comment:
Is "closed" the correct term? "Opened" for generic types could refer to `Type<>`. I didn't think that `Type<T>` would be considered "open".

Reply:
I think so: https://github.com/dotnet/csharplang/blob/main/spec/types.md#open-and-closed-types

I think `Type<>` is an *unbound* generic type while `Type<T>` is an *open* generic type.